airbyte-cdk/python/airbyte_cdk/sources/declarative/low_code_component_schema.yaml

```diff
@@ -0,0 +1,988 @@
+        type: string
+      $options:
+        type: object
+        additionalProperties: true
+  CustomStreamSlicer:
+    description: Stream slicer component whose behavior is derived from a custom code implementation of the connector
+    type: object
+    additionalProperties: true
+    required:
+      - type
+      - class_name
+    properties:
+      type:
+        type: string
+        enum: [CustomStreamSlicer]
+      class_name:
+        type: string
+      $options:
+        type: object
+        additionalProperties: true
+  DatetimeStreamSlicer:
+    description: Stream slicer that slices the stream over a datetime range
+    type: object
+    required:
+      - type
+      - cursor_field
+      - end_datetime
+      - datetime_format
+      - start_datetime
+      - step
+    properties:
+      type:
+        type: string
+        enum: [DatetimeStreamSlicer]
+      cursor_field:
+        type: string
+      datetime_format:
+        type: string
+      end_datetime:
+        anyOf:
+          - type: string
+          - "$ref": "#/definitions/MinMaxDatetime"
+      start_datetime:
+        anyOf:
+          - type: string
+          - "$ref": "#/definitions/MinMaxDatetime"
+      step:
+        type: string
+      end_time_option:
+        "$ref": "#/definitions/RequestOption"
+      lookback_window:
+        type: string
+      start_time_option:
+        "$ref": "#/definitions/RequestOption"
+      stream_state_field_end:
+        type: string
+      stream_state_field_start:
+        type: string
+      $options:
+        type: object
+        additionalProperties: true
+  DeclarativeOauth2Authenticator:
+    description: Authenticator for requests using Oauth authentication
+    type: object
+    required:
+      - type
+      - client_id
+      - client_secret
+      - refresh_token
+      - token_refresh_endpoint
+    properties:
+      type:
+        type: string
+        enum: [DeclarativeOauth2Authenticator]
+      client_id:
+        type: string
+      client_secret:
+        type: string
+      refresh_token:
+        type: string
+      token_refresh_endpoint:
+        type: string
+      access_token_name:
+        type: string
+      expires_in_name:
+        type: string
+      grant_type:
+        type: string
+      refresh_request_body:
+        type: object
+        additionalProperties: true
+      scopes:
+        type: array
+        items:
+          type: string
+      token_expiry_date:
+        type: string
+  DeclarativeStream:
+    description: A stream whose behavior is described by a set of declarative low code components
+    type: object
+    additionalProperties: true
+    required:
+      - type
+      - retriever
+    properties:
+      type:
+        type: string
+        enum: [DeclarativeStream]
+      retriever:
+        "$ref": "#/definitions/SimpleRetriever"
+      checkpoint_interval:
+        type: integer
+      name:
+        type: string
+        default: ""
+      primary_key:
```

Comment:
is it worth defining the primary key as a first class citizen since it's pretty complex?

Reply:
we could do the same with options instead of redefining it in every component

Reply:
Yup will move primary key to its own component and will try retesting. 

For options, even though options is used everywhere, it is still a pretty simplistic just a type object. I think it's slightly less intuitive to use a ref here because its just an extra hop to see what the actual object is. If it does get more complex, then we could go with the slight schema refactor down the road, but we should have some flexibility since this wouldn't be a breaking change to existing connectors. How does that sound to you?

Reply:
That’s a good plan!